packages/cli-lib/lib/handleFieldsJs.js

```diff
@@ -0,0 +1,218 @@
+
+/**
+ * FieldsJS Class.
+ * @param {string} projectDir - The root directory of the filePath
+ * @param {string} filePath - The path to the fields.js file to be converted
+ * @param {string} [rootWriteDir] - (Optional) The root of the directory in which to output the fields.js. If blank, a temporary directory is created.
+ */
+class FieldsJs {
+  constructor(projectDir, filePath, rootWriteDir) {
+    this.projectDir = projectDir;
+    this.filePath = filePath;
+
+    // Create tmpDir if no writeDir is given.
+    this.rootWriteDir =
+      rootWriteDir === undefined
+        ? createTmpDirSync('hubspot-temp-fieldsjs-output-')
+        : rootWriteDir;
+  }
+
+  /**
+   * Converts a fields.js file into a fields.json file, writes, and returns of fields.json
+   * @param {string} file - The path of the fields.js javascript file.
+   * @param {string} writeDir - The directory to write the file to.
+   * @returns {Promise} finalPath - Promise that returns path of the written fields.json file.
+   */
+  convertFieldsJs(writeDir) {
+    const filePath = this.filePath;
+    const options = yargs.argv.options;
+    const dirName = path.dirname(filePath);
+
+    logger.info(
+      i18n(`${i18nKey}.converting`, {
+        src: dirName + '/fields.js',
+        dest: dirName + '/fields.json',
+      })
+    );
+
+    try {
+      /* Since we require() fields.js files, node will cache them.
+       * Thus during an hs watch, if the user edits a fields.js file, node will return the cached version on require() instead of re-requiring the new one.
+       * So we clean the cache before requiring so that node requires the fresh fields.js. See https://nodejs.org/api/modules.html#caching.
+       */
+      delete require.cache[filePath];
+
+      // Save CWD and then switch CWD to the project. This is so that any calls to the file system that are written relatively will resolve properly.
+      const cwd = getCwd();
+      process.chdir(dirName);
+      /*
+       * If the dev marks their exported function as async, then require(filePath) returns an async function. In that case, fieldsArray is going to be a Promise.
+       * Further, it is expected that devs use await on any asyncronous calls.
+       * But fieldsArray _might_ not be a Promise. In order to be sure that it is, we use Promise.resolve.
+       */
+      const fieldsArray = require(filePath)(options);
+      return Promise.resolve(fieldsArray).then(fields => {
+        if (!Array.isArray(fields)) {
+          throw new SyntaxError(`${filePath} does not return an array.`);
+        }
+
+        let finalPath = path.join(writeDir, '/fields.json');
+        let json = fieldsArrayToJson(fields);
+        fs.outputFileSync(finalPath, json);
+        logger.info(
+          i18n(`${i18nKey}.converted`, {
+            src: dirName + '/fields.js',
+            dest: dirName + '/fields.json',
+          })
+        );
+        // Switch back to the original directory.
+        process.chdir(cwd);
```

Comment:
If this promise is rejected, do we still need to change the dir back?

Reply:
Some slack conversations were had but just to summarize the change I ended up making:

We don't want to kill the process if there is an issue with the `fields.js`. This makes it so that if `hs watch` is running and the user makes a mistake in their fields.js, the process continues. The DevEx is definitely better this way. Further, if it trips up on `hs upload`, the desired behaviour will be to log that an error occurred and continue uploading. This also mimics how `hs watch` currently runs - it will log an error and not upload, but keep watching.

So the way this is done is by adding a `resolved` prop to the FieldsJs object. Then, the process to the FieldsJs we:
1. Make a new FieldsJs object pointing to the correct file path
2. `await` the `.getOutputPathPromise()` method of the FieldsJs object.
3. We check the `resolved` prop. If this promise failed then this will be true. If it's true, don't upload (or in the case of `getFilesByType()`, don't add it to the filePath arrays).
4. If it's false, then set the `outputPath` prop of the fieldsjs object to the return value.

The reason I had to change `getFilesByType()` is so that we can stop errored `fields.js` files before they get added to the arrays, otherwise we'd need to figure out how to remove/skip the rejected Promises during the upload.